x/evm/keeper/state_transition.go

```diff
@@ -231,8 +231,9 @@ func (k *Keeper) ApplyMessage(evm *vm.EVM, msg core.Message, cfg *params.ChainCo
 	}
 	// Should check again even if it is checked on Ante Handler, because eth_call don't go through Ante Handler.
 	if msg.Gas() < intrinsicGas {
-		// eth_estimateGas will check for this exact error
-		return nil, stacktrace.Propagate(core.ErrIntrinsicGas, "intrinsic gas too low")
+		// eth_estimateGas will check for this exact error,
+		// don't wrap it with stacktrace, that makes it very hard to decipher for the client
+		return nil, core.ErrIntrinsicGas
```

Comment:
maybe consider using `error.Is`?

Reply:
I think we can’t do that because it get serialized to string and transfer to the client.

Reply:
Actually I've tried another approach which put the binary search logic into the grpc handler, creating a `EstimateGas` grpc endpoint.

Reply:
Just done that, looks better now.

Reply:
> Actually I've tried another approach which put the binary search logic into the grpc handler, creating a EstimateGas grpc endpoint. But at that time I don't how to make each iteration run in a clean context, maybe I can try CacheContext.

I think we should be using the binary search on the client-side instead. It might not be as clean but at least we move unnecessary computation from the grpc handler. Thoughts? cc: @khoslaventures

Reply:
In a default setup both rpc apis and grpc apis are kind of server side?

Reply:
And if one set log level to info (I think that’s the default), there’s a log for each grpc query, it would be confusing if one rpc call emites several grpc query logs.